File: rust/core/tests/ast_test.rs
//! Port of `AstTest.java`. Java's defensive-copy/`UnsupportedOperationException`
//! clauses are dropped — Rust's owned `Vec` fields are immutable by construction
//! — leaving each test's constructor + accessor core. The two reflection tests
//! (`blockPermitsExactlySevenVariants` / `tableOrFencePermitsExactlyTableAndFence`)
//! are dropped: the Rust enums *are* the compiler-enforced closed sets.

use varar_core::ast::{
    Block, Blockquote, Example, Fence, Heading, ListItem, Paragraph, Row, SegmentOffset, Table,
    TableOrFence, ThematicBreak, VarDoc,
};
use varar_core::span::Span;

const SPAN: Span = Span {
    start_offset: 0,
    end_offset: 5,
    start_line: 1,
    start_col: 1,
    end_line: 1,
    end_col: 6,
};

#[test]
fn segment_offset_exposes_both_offsets() {
    let offset = SegmentOffset::new(3, 7);
    assert_eq!(3, offset.text_offset);
    assert_eq!(7, offset.source_offset);
}

#[test]
fn heading_exposes_level_text_and_span_and_is_a_block() {
    let heading = Heading {
        level: 2,
        text: "Title".to_string(),
        span: SPAN,
    };
    assert_eq!(2, heading.level);
    assert_eq!("Title", heading.text);
    assert_eq!(SPAN, heading.span);
    let _: Block = Block::Heading(heading);
}

#[test]
fn paragraph_exposes_fields_and_is_a_block() {
    let paragraph = Paragraph {
        text: "Some text.".to_string(),
        span: SPAN,
        segment_map: vec![SegmentOffset::new(0, 0)],
    };
    assert_eq!("Some text.", paragraph.text);
    assert_eq!(SPAN, paragraph.span);
    assert_eq!(1, paragraph.segment_map.len());
    let _: Block = Block::Paragraph(paragraph);
}

#[test]
fn list_item_exposes_fields() {
    let marker_span = Span {
        start_offset: 0,
        end_offset: 2,
        start_line: 1,
        start_col: 1,
        end_line: 1,
        end_col: 3,
    };
    let list_item = ListItem {
        text: "An item".to_string(),
        span: SPAN,
        segment_map: vec![SegmentOffset::new(0, 0)],
        ordered: true,
        marker_span,
    };
    assert_eq!("An item", list_item.text);
    assert_eq!(SPAN, list_item.span);
    assert!(list_item.ordered);
    assert_eq!(marker_span, list_item.marker_span);
    let _: Block = Block::ListItem(list_item);
}

#[test]
fn blockquote_exposes_fields() {
    let blockquote = Blockquote {
        text: "Quoted".to_string(),
        span: SPAN,
        segment_map: vec![SegmentOffset::new(0, 0)],
    };
    assert_eq!("Quoted", blockquote.text);
    assert_eq!(SPAN, blockquote.span);
    let _: Block = Block::Blockquote(blockquote);
}

#[test]
fn row_exposes_fields() {
    let row = Row {
        cells: vec!["a".to_string(), "b".to_string()],
        cell_spans: vec![SPAN, SPAN],
        span: SPAN,
    };
    assert_eq!(vec!["a".to_string(), "b".to_string()], row.cells);
    assert_eq!(vec![SPAN, SPAN], row.cell_spans);
    assert_eq!(SPAN, row.span);
}

#[test]
fn table_exposes_fields() {
    let header = Row {
        cells: vec!["h1".to_string(), "h2".to_string()],
        cell_spans: vec![SPAN, SPAN],
        span: SPAN,
    };
    let data_row = Row {
        cells: vec!["v1".to_string(), "v2".to_string()],
        cell_spans: vec![SPAN, SPAN],
        span: SPAN,
    };
    let table = Table {
        span: SPAN,
        header: header.clone(),
        rows: vec![data_row],
    };
    assert_eq!(SPAN, table.span);
    assert_eq!(header, table.header);
    assert_eq!(1, table.rows.len());
    let _: Block = Block::Table(table);
}

#[test]
fn fence_exposes_fields() {
    let body_span = Span {
        start_offset: 1,
        end_offset: 4,
        start_line: 1,
        start_col: 2,
        end_line: 1,
        end_col: 5,
    };
    let fence = Fence {
        span: SPAN,
        info: "json".to_string(),
        body: "{}".to_string(),
        body_span,
    };
    assert_eq!(SPAN, fence.span);
    assert_eq!("json", fence.info);
    assert_eq!("{}", fence.body);
    assert_eq!(body_span, fence.body_span);
    let _: Block = Block::Fence(fence);
}

#[test]
fn thematic_break_exposes_span() {
    let thematic_break = ThematicBreak { span: SPAN };
    assert_eq!(SPAN, thematic_break.span);
    let _: Block = Block::ThematicBreak(thematic_break);
}

#[test]
fn example_exposes_fields() {
    let example = Example {
        scope_stack: vec!["Feature".to_string(), "Scenario".to_string()],
        span: SPAN,
        body: vec![Block::ThematicBreak(ThematicBreak { span: SPAN })],
    };
    assert_eq!(
        vec!["Feature".to_string(), "Scenario".to_string()],
        example.scope_stack
    );
    assert_eq!(SPAN, example.span);
    assert_eq!(1, example.body.len());
}

#[test]
fn var_doc_exposes_fields() {
    let example = Example {
        scope_stack: vec![],
        span: SPAN,
        body: vec![Block::ThematicBreak(ThematicBreak { span: SPAN })],
    };
    let orphan = TableOrFence::Fence(Fence {
        span: SPAN,
        info: String::new(),
        body: String::new(),
        body_span: SPAN,
    });
    let doc = VarDoc {
        path: "spec.md".to_string(),
        source: "# Title".to_string(),
        examples: vec![example],
        orphan_attachments: vec![orphan],
    };
    assert_eq!("spec.md", doc.path);
    assert_eq!("# Title", doc.source);
    assert_eq!(1, doc.examples.len());
    assert_eq!(1, doc.orphan_attachments.len());
}

File: java/core/src/test/java/dev/varar/core/AstTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.Test;

/** Port of typescript/packages/core/src/ast.ts (type definitions only, no logic there). */
class AstTest {

    private static final Span SPAN = new Span(0, 5, 1, 1, 1, 6);

    @Test
    void segmentOffsetExposesBothOffsets() {
        Ast.SegmentOffset offset = new Ast.SegmentOffset(3, 7);
        assertEquals(3, offset.textOffset());
        assertEquals(7, offset.sourceOffset());
    }

    @Test
    void headingExposesLevelTextAndSpanAndIsABlock() {
        Ast.Heading heading = new Ast.Heading(2, "Title", SPAN);
        assertEquals(2, heading.level());
        assertEquals("Title", heading.text());
        assertEquals(SPAN, heading.span());
        assertTrue(heading instanceof Ast.Block);
    }

    @Test
    void paragraphExposesFieldsAndDefensivelyCopiesSegmentMap() {
        List<Ast.SegmentOffset> mutable = new ArrayList<>();
        mutable.add(new Ast.SegmentOffset(0, 0));
        Ast.Paragraph paragraph = new Ast.Paragraph("Some text.", SPAN, mutable);

        assertEquals("Some text.", paragraph.text());
        assertEquals(SPAN, paragraph.span());
        assertEquals(1, paragraph.segmentMap().size());

        // Mutating the caller's list after construction must not affect the record.
        mutable.add(new Ast.SegmentOffset(1, 1));
        assertEquals(1, paragraph.segmentMap().size());

        // The record's own list must be unmodifiable.
        assertThrows(
                UnsupportedOperationException.class,
                () -> paragraph.segmentMap().add(new Ast.SegmentOffset(2, 2)));
        assertTrue(paragraph instanceof Ast.Block);
    }

    @Test
    void listItemExposesFieldsAndDefensivelyCopiesSegmentMap() {
        List<Ast.SegmentOffset> mutable = new ArrayList<>(List.of(new Ast.SegmentOffset(0, 0)));
        Span markerSpan = new Span(0, 2, 1, 1, 1, 3);
        Ast.ListItem listItem = new Ast.ListItem("An item", SPAN, mutable, true, markerSpan);

        assertEquals("An item", listItem.text());
        assertEquals(SPAN, listItem.span());
        assertTrue(listItem.ordered());
        assertEquals(markerSpan, listItem.markerSpan());
        assertThrows(
                UnsupportedOperationException.class, () -> listItem.segmentMap().add(new Ast.SegmentOffset(1, 1)));
        assertTrue(listItem instanceof Ast.Block);
    }

    @Test
    void blockquoteExposesFieldsAndDefensivelyCopiesSegmentMap() {
        List<Ast.SegmentOffset> mutable = new ArrayList<>(List.of(new Ast.SegmentOffset(0, 0)));
        Ast.Blockquote blockquote = new Ast.Blockquote("Quoted", SPAN, mutable);

        assertEquals("Quoted", blockquote.text());
        assertEquals(SPAN, blockquote.span());
        assertThrows(
                UnsupportedOperationException.class,
                () -> blockquote.segmentMap().add(new Ast.SegmentOffset(1, 1)));
        assertTrue(blockquote instanceof Ast.Block);
    }

    @Test
    void rowExposesFieldsAndDefensivelyCopiesCellsAndCellSpans() {
        List<String> cells = new ArrayList<>(List.of("a", "b"));
        List<Span> cellSpans = new ArrayList<>(List.of(SPAN, SPAN));
        Ast.Row row = new Ast.Row(cells, cellSpans, SPAN);

        assertEquals(List.of("a", "b"), row.cells());
        assertEquals(List.of(SPAN, SPAN), row.cellSpans());
        assertEquals(SPAN, row.span());
        assertThrows(UnsupportedOperationException.class, () -> row.cells().add("c"));
        assertThrows(UnsupportedOperationException.class, () -> row.cellSpans().add(SPAN));
    }

    @Test
    void tableExposesFieldsAndDefensivelyCopiesRows() {
        Ast.Row header = new Ast.Row(List.of("h1", "h2"), List.of(SPAN, SPAN), SPAN);
        Ast.Row dataRow = new Ast.Row(List.of("v1", "v2"), List.of(SPAN, SPAN), SPAN);
        List<Ast.Row> rows = new ArrayList<>(List.of(dataRow));
        Ast.Table table = new Ast.Table(SPAN, header, rows);

        assertEquals(SPAN, table.span());
        assertEquals(header, table.header());
        assertEquals(1, table.rows().size());
        assertThrows(UnsupportedOperationException.class, () -> table.rows().add(dataRow));
        assertTrue(table instanceof Ast.Block);
    }

    @Test
    void fenceExposesFields() {
        Span bodySpan = new Span(1, 4, 1, 2, 1, 5);
        Ast.Fence fence = new Ast.Fence(SPAN, "json", "{}", bodySpan);

        assertEquals(SPAN, fence.span());
        assertEquals("json", fence.info());
        assertEquals("{}", fence.body());
        assertEquals(bodySpan, fence.bodySpan());
        assertTrue(fence instanceof Ast.Block);
    }

    @Test
    void thematicBreakExposesSpan() {
        Ast.ThematicBreak thematicBreak = new Ast.ThematicBreak(SPAN);
        assertEquals(SPAN, thematicBreak.span());
        assertTrue(thematicBreak instanceof Ast.Block);
    }

    @Test
    void exampleExposesFieldsAndDefensivelyCopiesScopeStackAndBody() {
        List<String> scopeStack = new ArrayList<>(List.of("Feature", "Scenario"));
        List<Ast.Block> body = new ArrayList<>(List.of(new Ast.ThematicBreak(SPAN)));
        Ast.Example example = new Ast.Example(scopeStack, SPAN, body, true);

        assertEquals(List.of("Feature", "Scenario"), example.scopeStack());
        assertEquals(SPAN, example.span());
        assertTrue(example.precededByDelimiter());
        assertEquals(1, example.body().size());
        assertThrows(
                UnsupportedOperationException.class, () -> example.scopeStack().add("Nope"));
        assertThrows(UnsupportedOperationException.class, () -> example.body().add(new Ast.ThematicBreak(SPAN)));
    }

    @Test
    void varDocExposesFieldsAndDefensivelyCopiesExamplesAndOrphanAttachments() {
        Ast.Example example = new Ast.Example(List.of(), SPAN, List.of(new Ast.ThematicBreak(SPAN)), true);
        List<Ast.Example> examples = new ArrayList<>(List.of(example));
        List<Ast.TableOrFence> orphanAttachments = new ArrayList<>(List.of(new Ast.Fence(SPAN, "", "", SPAN)));
        Ast.VarDoc doc = new Ast.VarDoc("oath.md", "# Title", examples, orphanAttachments);

        assertEquals("oath.md", doc.path());
        assertEquals("# Title", doc.source());
        assertEquals(1, doc.examples().size());
        assertEquals(1, doc.orphanAttachments().size());
        assertThrows(UnsupportedOperationException.class, () -> doc.examples().add(example));
        assertThrows(
                UnsupportedOperationException.class,
                () -> doc.orphanAttachments().add(new Ast.Fence(SPAN, "", "", SPAN)));
    }

    @Test
    void blockPermitsExactlySevenVariants() {
        assertEquals(7, Ast.Block.class.getPermittedSubclasses().length);
    }

    @Test
    void tableOrFencePermitsExactlyTableAndFence() {
        Class<?>[] permitted = Ast.TableOrFence.class.getPermittedSubclasses();
        assertEquals(2, permitted.length);
        assertEquals(Set.of(Ast.Table.class, Ast.Fence.class), Set.of(permitted));
    }
}
